### ISSUE TITLE
Add asynchronous interface for libuv itself

### ISSUE THREAD

**Issue by rmtew** [joyent/libuv#262]
Currently the control flow for the libuv framework is (as I understand it):

  ... set up application ...
  uv_run()
  ... return from blocking indefinitely ...

This  means that if an application is to use libuv, it needs to let libuv drive it.  This affects the way in which the application can be written.  If there was a minimal change to the libuv running mechanism to let it be pumped with a specified timeout, then an application could instead drive libuv more in the manner of a library and a main loop could be written in a straight-forward and synchronous manner.

I suggest that this could be done with the following changes:
1. The contents of the while loop in UV_RUN are moved to UV_PUMP.
2. A uv_pump() method is added that takes a optional timeout parameter.
3. The poll method observes the timeout parameter instead of the result of uv_get_poll_timeout(), or perhaps gets it from that function for this pending one-off iteration.

I can write and provide a patch, if there is an indication that one would be applied.

Cheers.

**Comment by bnoordhuis** [comment]
Sure, I'd take a patch for that. I don't think libev has facilities for timed polls (without adding explicit timers to the event loop) but the diff below should be all you need for a simple immediate poll.

``` diff
diff --git a/src/unix/core.c b/src/unix/core.c
index cf3138f..2cf0ee5 100644
--- a/src/unix/core.c
+++ b/src/unix/core.c
@@ -194,6 +194,12 @@ int uv_run(uv_loop_t* loop) {
 }


+int uv_poll(uv_loop_t* loop) {
+  ev_run(loop->ev, EVRUN_NOWAIT);
+  return 0;
+}
+
+
 void uv__handle_init(uv_loop_t* loop, uv_handle_t* handle,
     uv_handle_type type) {
   loop->counters.handle_init++;
```

**Comment by txdv** [comment]
looks like a run_once method to me, now the windows equivalent is the only thing missing.

**Comment by mattn** [comment]
I hope `uv_run_once()`.

**Comment by bnoordhuis** [comment]
I called it `uv_poll()` because it lets you poll for events at your convenience but the actual name is inconsequential. If someone wants to submit a patch that includes Windows support, please go ahead, else it will have to wait until someone on the team has some spare cycles.

**Comment by mattn** [comment]
Do you mean that the `uv_poll()` do once of `uv_run()`'s loop?

**Comment by bnoordhuis** [comment]
Yes, poll once without blocking if there are no events.

**Comment by mattn** [comment]
But `uv_poll` is defined in src/core/win.c alreadly. Do you mean changing `uv_poll` to global function?

**Comment by bnoordhuis** [comment]
Oh, I see what you mean. Like I said, the name is inconsequential - if the Windows component of libuv already defines a `uv_poll()`, we'll name it `uv_run_once()`.

**Comment by bnoordhuis** [comment]
Fixed in e53cecb.
